### PR TITLE
Add babel/es6 and react to client build

### PR DESCRIPTION
**Pull request by carlfeberhard**
James asked me to PR this to get the ball rolling towards react (edit: see also #2451).
# Babel

This (1) adds babel's es6 parsing (https://babeljs.io/) to the client build. This is via webpack only at the moment. There are some difficulties in implementing es6 for scripts not loaded by webpack that I'll describe in a comment below.

Practically, that means anything specified as a webpack entry point or any modules imported into one of those entry points can use es6:
https://github.com/galaxyproject/galaxy/blob/dev/client/webpack.config.js#L37

I'd recommend creating smaller apps to cover existing sections of Galaxy and then attempting to bring them together (if that's even necessary) as opposed to expanding the analysis app, but I know there are differing opinions on the team. (As an example: https://github.com/carlfeberhard/galaxy/tree/history.app/client/galaxy/scripts/apps)

A few notes:
- Existing scripts loaded via other means **can't use es6** with this PR. However, es5 is still valid es6 so you can import Backbone models, etc. **into** es6 apps and reuse code.
- I explored react.backbone and other combinations thereof and, while absolutely possible to connect react renderers to backbone models and collections, it seems almost equally as much work to move directly to flux or redux

Both stage-0 and decorators have been added to the `.babelrc`:
- stage-0 allows most es6 features to be used (https://babeljs.io/docs/plugins/preset-stage-0/) 
- and decorators will allow redux's connect and the convenient `@autobind` decorator if you want to add those later. Note, though, that nothing in _this_ PR requires that plugin, so feel free to remove it if need be. I've found it useful. (https://github.com/loganfsmyth/babel-plugin-transform-decorators-legacy)
# React

This also (2) adds react (https://facebook.github.io/react/index.html) and react-dom to the libraries (and react preset to babel). This allows full JSX, and es6. Since I'm unaware whether the team wants to go with Jest or Enzyme, neither were included and I'd encourage some talk about that.

I've also specifically left out both flux or redux as that's a decision the team should make as well before moving forward. Both is a heavy shift from Backbone, but flux's multiple stores might be less of a mental shift than redux's immutability and reducers. I've always bet that, due to it's size, Galaxy would most benefit from redux's straightforward state update system, so I'd think you'd want to move towards that sooner rather than later.

### CODE REVIEW

**Comment by carlfeberhard** [comment]
_So, Why no ES6 for grunt/non-webpack scripts?_
There are two issues with this: sourcemaps and uglification.
- a chain of sources (a second sourcemap pointing to locations in a first sourcemap which points to locations in an un-minified source file) is theoretically possible, but I couldn't get it to work. IIRC, it was either an issue with uglify's sourcemap generation or the way we place/serve sourcemaps
- uglify doesn't understand es6 and can't minify it directly. You'd need to transform es6->es5 then uglify it (prob. in place). I attempted https://www.npmjs.com/package/babel-plugin-uglify but couldn't get it to work. It may be best to persue babili (https://babeljs.io/blog/2016/08/30/babili) even though it's only beta. Babel's built-in compression only removes whitespace.

File: client/galaxy/scripts/apps/history-panel.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 var RightPanel = require( 'layout/panel' ).RightPanel,
     Ui = require( 'mvc/ui/ui-misc' ),
-    historyOptionsMenu = require( 'mvc/history/options-menu' );
+    historyOptionsMenu = require( 'mvc/history/options-menu' ),
```
Comment (martenson) [inline client/galaxy/scripts/apps/history-panel.js:3]: seems like a dev remnant?

Comment (carlfeberhard) [reply client/galaxy/scripts/apps/history-panel.js:3]: Nah - it's a minor fix. Without that change, `CurrentHistoryView` and `_l` become globals.

Comment (martenson) [reply client/galaxy/scripts/apps/history-panel.js:3]: damn, I am so unused to that multi var syntax, sorry :D

Comment (carlfeberhard) [reply client/galaxy/scripts/apps/history-panel.js:3]: No worries - I appreciate the review!

**Comment by dannon** [comment]
I tried a local merge of this (here at https://github.com/dannon/galaxy/tree/test_merge_react) but there were a few problems I ran into with resources not being found.  Haven't had a chance to really dig into _what_ yet, but it might be something in the require logic isn't quite right.  Will work on it more, but just FYI in case anyone else is attempting to test this.

**Comment by jmchilton** [comment]
Squint tightly and a version of this has been integrated, closing this now.

**Pull request closed without merging**